### PR TITLE
Use SOURCE_DATA_EPOCH for reproducible builds

### PR DESCRIPTION
**Pull request by richsalz**
Thanks to Juro Bystricky for the suggestion and prototype.
Fixes: #4639 

I don't know if this is a bugfix and worth putting to 1.1.0 or not.

### CODE REVIEW

File: util/mkbuildinf.pl [hunk 1]
```diff
@@ -1,26 +1,45 @@
...
+use warnings;
 
 my ($cflags, $platform) = @ARGV;
-
 $cflags = "compiler: $cflags";
-$date = localtime();
+
+my $date = "built on: " . localtime();
+$date = "reproducible build, date unspecified"
+    if defined $ENV{'SOURCE_DATE_EPOCH'};
```
Comment (kroeckx) [inline util/mkbuildinf.pl]: The point of SOURCE_DATE_EPOCH is that you actually use that date.

File: crypto/cversion.c [hunk 2]
```diff
@@ -18,41 +18,22 @@ unsigned long OpenSSL_version_num(void)
...
-#ifdef CFLAGS
-        return CFLAGS;
-#else
-        return "compiler: information not available";
-#endif
-    }
-    if (t == OPENSSL_PLATFORM) {
-#ifdef PLATFORM
+    case OPENSSL_CFLAGS:
+        return compiler_flags;;
```
Comment (kroeckx) [inline crypto/cversion.c]: 1 ;

**Comment by richsalz** [comment]
Additional commit pushed (to be squashed) that addresses Kurt's issues.

File: util/mkbuildinf.pl [hunk 3]
```diff
@@ -1,26 +1,43 @@
...
 
+use strict;
+use warnings;
 
 my ($cflags, $platform) = @ARGV;
-
 $cflags = "compiler: $cflags";
-$date = localtime();
+
+my $date = $ENV{'SOURCE_DATE_EPOCH'} || localtime();
```
Comment (levitte) [inline util/mkbuildinf.pl]: `$ENV{'SOURCE_DATE_EPOCH'}` is the epoch, i.e. just a simple integer, which should be fed to `localtime`, so this is what that assignment should look like:

``` perl
$date = localtime($ENV{'SOURCE_DATE_EPOCH'} // time());
```

Also, we might want to consider `gmtime` over `localtime` and make sure we add `" UTC"` to the date.

Comment (richsalz) [reply util/mkbuildinf.pl]: fixed, and converted to gmtime; thanks.  additional commit pushed.

**Comment by levitte** [comment]
Hmmm, it's a pity we don't have `./util/shlib_wrap.sh apps/openssl version -a` in `.travis.yml`, that would immediately show us....

Actually, come to think  of it, you *could* write a small test.  I have an idea, would you like some help?

**Comment by richsalz** [comment]
@levitte -- I don't understand your last comment?

**Comment by richsalz** [comment]
ping anyone.

**Comment by richsalz** [comment]
What do you think about picking to 1.1.0?  It didn't go cleanly, but the doing it is simple.  I'm inclined to not do it unless you think I should.

**Pull request closed without merging**